internal/oidc/token/token_handler.go

```diff
@@ -320,6 +346,14 @@ func upstreamLDAPRefresh(ctx context.Context, providerCache oidc.UpstreamIdentit
 	// Replace the old value with the new value.
 	session.Fosite.Claims.Extra[oidc.DownstreamGroupsClaim] = groups
 
+	added, removed := diffSortedGroups(oldGroups, groups)
+	if len(added) > 0 {
+		warning.AddWarning(ctx, "", fmt.Sprintf("User %q has been added to the following groups: %q", username, added))
+	}
+	if len(removed) > 0 {
+		warning.AddWarning(ctx, "", fmt.Sprintf("User %q has been removed from the following groups: %q: ", username, removed))
+	}
```

Comment:
re:

> Can we implement the warning in a generic way / at a high level so that we do not need to remember to add them for each new upstream refresh?

At a minimum have a shared `warnIfGroupsChanged(ctx, username, newGroups, oldGroups)` function that encapsulates this duplicated logic.